# 1.- California Housing

En esta sección, crearemos un modelo de red neuronal para predecir los precios de viviendas dada una serie de características, haciendo de este un problema de regresión. El fin de esta tarea es jugar con diferentes arquitecturas y herramientas de redes neuronales, con lo que no se pide hacer un análisis exploratorio de los datos ni entenderlos, aunque eso queda a elección del alumno.

## Imports

Importamos las librerías necesarias

In [45]:
import numpy as np
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

## Importamos el dataset y lo preprocesamos

Importamos el dataset, que está disponible en la librería de scikitlearn. Tras ello, separamos en train y test y estandarizamos los datos - recordad aquí que las redes neuronales funcionan mejor con valores centrados en 0.

*Nota: los valores de las etiquetas no los normalizamos porque ya vienen normalizados en el propio dataset.*

In [46]:
# Cargar el conjunto de datos de California Housing
housing_data = fetch_california_housing()

# Separar características y etiquetas
X = housing_data.data
y = housing_data.target

# Dividir el conjunto de datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Escalar características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

## Implementación y entrenamiento de la red neuronal

Pasamos aquí a diseñar y entrenar nuestra red neuronal. Alguna de las cosas que podemos probar es:

- Número de capas y neuronas, así como capas de drop out o algún otro tipo de regularización al crear la red neuronal con el modelo *Sequential*. ¡OJO! Tenemos que pensar, especialmente, en qué función de activación y tamaño queremos en la última capa.
- Optimizador al compilar el modelo. Además aquí la función de perdida será el MSE, ya que es la métrica a optimizar, como está explicado en el documento de texto de la tarea.
- Número de epochs o tamaño del batch, que le podemos indicar a la red en el entrenamiento - método *fit*.

Estas directrices valen para el otro dataset que usaremos en la siguiente sección.

*Nota: se recomienda empezar con una red sencilla, de una capa y unas pocas neuronas, e ir añadiendo complejidad poco a poco. No empezemos con muchas capas y técnicas de regularización de primeras, sobre todo si no es necesario.*

In [47]:
# Definir el modelo de la red neuronal con Keras
model = Sequential()

# Añadir una capa oculta con 10 neuronas y activación ReLU
model.add(Dense(10, activation='relu', input_shape=(X_train_scaled.shape[1],)))

# Añadir la capa de salida con 1 neurona (regresión → salida continua)
model.add(Dense(1, activation='linear'))

# Compilar el modelo
model.compile(loss='mean_squared_error', optimizer = 'adam')

# Entrenar el modelo
model.fit(X_train_scaled, y_train, epochs=100, batch_size=32, verbose=1)

# Evaluar el modelo en el conjunto de prueba
mse = model.evaluate(X_test_scaled, y_test)

print("Error cuadrático medio en el conjunto de prueba:", mse)

Epoch 1/100


c:\Users\nunoc\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - loss: 3.2068
Epoch 2/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - loss: 0.8059
Epoch 3/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - loss: 0.6332
Epoch 4/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - loss: 0.5537
Epoch 5/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - loss: 0.4807
Epoch 6/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - loss: 0.4463
Epoch 7/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - loss: 0.4322
Epoch 8/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - loss: 0.4165
Epoch 9/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - loss: 0.4260
Epoch 10/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - loss: 0.4135
Epoch 11/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.4047
Epoch 12/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.4047
Epoch 13/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - loss: 0.4088
Epoch 14/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - loss: 0.4053
Epoch 15/100
516/516 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step -

En este primer ejercicio, entrené una red neuronal simple para resolver un problema de regresión: predecir el precio de viviendas en California. Empecé con una arquitectura básica de solo una capa oculta con 10 neuronas y función de activación ReLU. La capa de salida tiene una única neurona con activación lineal, ya que estamos prediciendo un valor continuo.

Probé con 100 épocas y un batch size de 32, usando MSE como función de pérdida. El modelo fue capaz de aprender bastante bien, reduciendo la pérdida a lo largo del entrenamiento. Al evaluar sobre el conjunto de prueba, obtuve un error cuadrático medio (MSE) de aproximadamente 0.34, lo cual considero un resultado razonable para una red tan sencilla.

# 2.- Clasificación de prendas de ropa - Fashion MNIST

Ahora, tenemos un dataset muy similar al de MNIST - las imágenes tienen el mismo tamaño -, pero con la diferencia de que ahora representan diferentes prendas de ropa que debemos clasificar.

## Imports

Importamos las librerías necesarias (algunas están repetidas por si se quiere copiar directamente este ejemplo en el futuro).

In [48]:
import numpy as np
from tensorflow.keras.datasets import fashion_mnist
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

## Importamos el dataset y lo preprocesamos

Importamos el dataset, que está disponible en la librería de keras. En este caso, ya viene separado por X_train, y_train, X_test e y_test. Después, normalizamos. Aquí, estamos normalizando para que los píxeles vayan entre 0 y 1, pero podríamos probar a estandarizar, como en el caso anterior. Esto no es necesario y se deja a elección del alumno.

Después, adaptamos la forma para que pase de ser una matriz bidimensional a un vector unidimensional.

In [49]:
# Cargar el conjunto de datos Fashion MNIST
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

# Normalizar las imágenes
X_train_normalized = X_train.astype('float32') / 255.0
X_test_normalized = X_test.astype('float32') / 255.0

# Redimensionar las imágenes a un vector unidimensional
X_train_flattened = X_train_normalized.reshape((X_train_normalized.shape[0], -1))
X_test_flattened = X_test_normalized.reshape((X_test_normalized.shape[0], -1))

## Implementación y entrenamiento de la red neuronal

Igual que antes. Ahora estamos con un problema de clasificación, con lo que tendremos que saber qué capa utilizar a la salida de la red, y de qué tamaño.

In [50]:
# Definir el modelo de la red neuronal con Keras
model = Sequential()

# Añadir una capa oculta con 128 neuronas y activación ReLU
model.add(Dense(128, activation='relu', input_shape=(X_train_flattened.shape[1],)))

# Añadir la capa de salida con 10 neuronas (una por clase)
model.add(Dense(10, activation='softmax'))

# Compilar el modelo
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# Entrenar el modelo
model.fit(X_train_flattened, y_train, epochs=10, batch_size=32, verbose=1)

# Evaluar el modelo en el conjunto de prueba
loss, accuracy = model.evaluate(X_test_flattened, y_test)

print("Pérdida en el conjunto de prueba:", loss)
print("Precisión en el conjunto de prueba:", accuracy)

Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.7838 - loss: 0.6269
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8640 - loss: 0.3825
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8762 - loss: 0.3392
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8809 - loss: 0.3206
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.8903 - loss: 0.2992
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8973 - loss: 0.2800
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9002 - loss: 0.2678
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9038 - loss: 0.2561
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.9058 - loss: 0.2481
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9119 - loss: 0.2385
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8858 - loss: 0.3257
Pérdida en el conjunto de prueba: 0.32390511

En este conjunto de datos 'Fashion MNIST', el objetivo era clasificar el tipo de pieza en cada imagen. Creé una red con un modelo simple que contenía una capa oculta de 128 neuronas con 'ReLu' y una capa de salida con 10 neuronas con 'softmax'. Además, normalicé los píxeles y aplané las imágenes, es decir, las convertí en vectores. Elegí la función de pérdida 'sparse_categorical_crossentropy' porque las etiquetas son números enteros del 0 al 9. El entrenamiento consistió en 10 épocas y un tamaño de lote de 32. En conclusión, la precisión en el conjunto de prueba fue de aproximadamente el 88%. Lo cual parece bien para una red simple. 